tc/autotuner/options_cache.h

```diff
@@ -0,0 +1,173 @@
+};
+
+/**
+ * A value in the options cache is exactly the content of the underlying proto
+ * in tc/proto/compcache.proto. An OptionsCacheValue is templated by the Backend
+ * and contains an object of the proper MappingOptions type as well as the
+ * runtimes of subsequent runs.
+ */
+template <typename Backend>
+struct OptionsCacheValue {
+  typename Backend::OptionsCacheValueProtoType toProtobuf() const;
+  static OptionsCacheValue<Backend> fromProtobuf(
+      const typename Backend::OptionsCacheValueProtoType& proto);
+
+  std::vector<Duration> runtimes;
+  typename Backend::MappingOptionsType mappingOptions;
+};
+
+/**
+ * An Options cache is a simple abstraction around an unordered_multimap of
+ * protobuf-backed key value pairs. It has an underlying store object and
+ * provides simple functions to load/store from proto, record, prune and
+ * extract topK values ordered by runtime.
+ * All store and retrieve operations are protected by taking a lock on an
+ * internal mutex and thus are threadsafe.
+ * An OptionsCache is templated by the backend type because the values stored
+ * are backend-dependent.
+ */
+template <typename Backend>
+struct OptionsCache {
+ public:
+  using KeyType = OptionsCacheKey;
+  using ValueType = OptionsCacheValue<Backend>;
+  using MultiMapType = std::unordered_multimap<
+      OptionsCacheKey,
+      OptionsCacheValue<Backend>,
+      OptionsCacheKeyHash>;
+
+  OptionsCache<Backend>() = default;
+  OptionsCache<Backend>(const OptionsCache<Backend>& other);
+
+  /// Clears the content of the cache and resets the counters
+  void clear();
+
+  /// \return the number of values for a particular key
+  size_t count(const OptionsCacheKey& key) const;
+
+  //// \return the number of elements in the cache
+  size_t size() const;
+
+  /// Collects the keys in the cache
+  /// \return an unordered_set of keys
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> getKeys() const;
+
+  /// Loads in place from proto file. Calls fromProto which can insert
+  /// duplicates, so be sure your cache is cleared if you don't want those
+  void loadCacheFromFile(const std::string& filename);
+
+  /// Stores to a proto file at the specified location
+  void storeCacheToFile(const std::string& filename) const;
+
+  /// Saves a new runtime.
+  /// If the key does not exist, a new entry is inserted.
+  /// If the key exists, a search is performed on options.
+  /// If the corresponding options are found, the duration is appended to the
+  /// runtimes, otherwise a new entry is inserted in the multimap.
+  void recordRuntime(
+      const lang::CanonicalTcString& tc,
+      const std::vector<TensorInfo>& inputs,
+      const std::vector<TensorInfo>& outputs,
+      const std::string& deviceStr,
+      const typename Backend::MappingOptionsType& options,
+      Duration duration);
+
+  /// Returns the top-K best mapping options for a particular
+  /// TC/inputs/outputs/device. Note that the result may be empty (in
```

Comment:
Looking at the implementation, it seems "best" is defined in terms of
the median run-time.  Perhaps this is obvious, by I wouldn't have guessed that.